staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,895 @@
+			state.mutex.Unlock()
+			ca.vniToVnStateMutex.Unlock()
+			return
+		}
+		// deletedAttRef is the last one in the virtual network state: proceed to remove the virtual netowrk state from the map
+		// storing all the states stop informer
+		delete(ca.vniToVnState, deletedAttRef.vni)
+		ca.vniToVnStateMutex.Unlock()
+		ca.stopInformerAndEnqueueRemoteAttRefs(state)
+		state.mutex.Unlock()
+		// TODO make these prints a single one
+		glog.V(5).Infof("Stopped informer on Remote attachments in Virtual Network with ID %d\n", deletedAttRef.vni)
+		glog.V(5).Infof("Deleted Virtual Network with ID %d state", deletedAttRef.vni)
+	} else {
+		glog.V(5).Infof("Removed local NetworkAttachment %s from Virtual Network with ID %d state", deletedAttRef.nsn, deletedAttRef.vni)
+	}
+}
+
+func (ca *ConnectionAgent) initRemoteAttsInformerAndLister(state *vnState, nsn k8stypes.NamespacedName, vni uint32) {
+	state.remoteAttsInformer, state.remoteAttsLister = v1a1AttsCustomNamespaceInformerAndLister(ca.kcs,
+		resyncPeriod,
+		nsn.Namespace,
+		fromFieldsSelectorToTweakListOptionsFunc(ca.remoteAttInVNWithVirtualIPAndHostIPSelector(vni)))
+	state.remoteAttsInformer.AddEventHandler(k8scache.ResourceEventHandlerFuncs{
+		AddFunc:    ca.onRemoteAttAdded,
+		UpdateFunc: ca.onRemoteAttUpdated,
+		DeleteFunc: ca.onRemoteAttRemoved})
+}
+
+func (ca *ConnectionAgent) onRemoteAttAdded(obj interface{}) {
+	remoteAtt := obj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment %#+v addition to remote NetworkAttachments cache for VNI %d\n", remoteAtt, remoteAtt.Spec.VNI)
+	remoteAttRef := ca.fromAttToEnquableRef(remoteAtt)
+	ca.queue.Add(remoteAttRef)
+}
+
+// TODO make the enqueuing conditional on a check on whether the update is relevant
+// (e.g. an update which adds a metadata label is completely irrelevant and should not
+// trigger an enqueuing)
+func (ca *ConnectionAgent) onRemoteAttUpdated(oldObj, newObj interface{}) {
+	oldRemoteAtt := oldObj.(*netv1a1.NetworkAttachment)
+	newRemoteAtt := newObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment update from %#+v to %#+v in remote NetworkAttachments cache\n", oldRemoteAtt, newRemoteAtt)
+	newRemoteAttRef := ca.fromAttToEnquableRef(newRemoteAtt)
+	ca.queue.Add(newRemoteAttRef)
+}
+
+func (ca *ConnectionAgent) onRemoteAttRemoved(obj interface{}) {
+	peeledObj := peel(obj)
+	remoteAtt := peeledObj.(*netv1a1.NetworkAttachment)
+	glog.V(5).Infof("Notified of NetworkAttachment %#+v removal from remote NetworkAttachments cache for VNI %d\n", remoteAtt, remoteAtt.Spec.VNI)
+	remoteAttRef := ca.fromAttToEnquableRef(remoteAtt)
+	ca.queue.Add(remoteAttRef)
+}
+
+func (ca *ConnectionAgent) startRemoteAttsInformer(state *vnState) {
+	state.remoteAttsInformerStopCh = make(chan struct{})
+	go state.remoteAttsInformer.Run(aggregateStopChannels(ca.stopCh, state.remoteAttsInformerStopCh))
+}
+
+func (ca *ConnectionAgent) getVNStateForVNI(vni uint32) (state *vnState, vnStateExists bool) {
```

Comment:
why return a bool that is redundant with `state != nil` ?

Reply:
If the result of the lookup must be used more than once in the function where `getVNStateForVNI` is invoked, I find it more readable to use a bool variable properly named rather than checking `state != nil` every time. But this is only a personal preference, it's objective that using your solution is cleaner/leaner. Also, if the user of the function prefers using a bool rather than checking `state != nil` every time it can build such a variable by himself after invoking `getVNStateForVNI` by assigning to such variale the result of the boolean expression `state != nil`.

Reply:
Yes, let the caller make itself a redundant boolean if it likes.

Reply:
Fixed in c95baba.